### PR TITLE
Shuffle blocks for `contentOnly` patterns

### PR DESCRIPTION
**Pull request by kevin940726**
## What?

A potential approach to https://github.com/WordPress/gutenberg/issues/44581/

> **Note**
> Code isn't optimized. Once we are fine with the approach, I'll try to polish it more.

## Why?

See https://github.com/WordPress/gutenberg/issues/44581

## How?

This PR implements an approach to swap non-content blocks with something else in other patterns. Since that inserted patterns don't have any information about which patterns they're inserted from, we have to perform an algorithm to "guess" which candidate patterns are alike. The algo used in this PR is to:

1. Flatten all blocks and inner blocks in the patterns.
2. Keep only the blocks that have `__experimentalRole="content"` in any of their attributes.
3. Compare the names of the flattened content blocks to the available patterns using the same method.
4. If both have the same length of content blocks then it's a match.

Once we find a match, we then add the "Shuffle" button to the toolbar. If the user clicks the button, it will randomly choose a pattern from the list and apply all but the contents to the blocks.

To prevent choosing the same pattern that it's added from, we do an `areBlocksAlike` check to see if the blocks look like existing patterns and filter it out.

## Testing Instructions

1. Insert a pattern.
2. Go to the Code Editor and add `"templateLock":"contentOnly"` to the top-most group of the inserted pattern.
3. Select the inserted pattern/group.
4. Click on the "Shuffle" button in the toolbar.
5. The pattern should be randomly changed to a different pattern.

## Screenshots or screencast 

https://user-images.githubusercontent.com/7753001/200731087-a9e3980b-08ca-4645-9e8b-94ce29330672.mp4

### CODE REVIEW

File: packages/block-editor/src/hooks/content-lock-ui.js [hunk 1]
```diff
@@ -41,6 +47,89 @@ function StopEditingAsBlocksOnOutsideSelect( {
...
+								( block, index ) =>
+									block.name ===
+									patternContentBlocks[ index ].name
+							)
+						);
+					} )
+					.filter(
+						( pattern ) =>
+							! areBlocksAlike( blocks, pattern.blocks )
+					),
```
Comment (Mamaduka) [inline packages/block-editor/src/hooks/content-lock-ui.js]: Have you tried using `__experimentalGetPatternsByBlockTypes` or `__experimentalGetPatternTransformItems`?

Comment (kevin940726) [reply packages/block-editor/src/hooks/content-lock-ui.js]: `__experimentalGetPatternsByBlockTypes` uses `blockTypes` in patterns, but unfortunately we don't have that information here.

`__experimentalGetPatternTransformItems` only handles blocks without inner blocks so we can't use that yet either.

**Comment by jameskoster** [comment]
Exciting! 

The Shuffle button is appearing, but I am seeing results that I would consider to be unexpected. For example I have a Group containing a single Paragraph block, and when I shuffle I'm presented with Header and Footer patterns. 

For the purpose of shuffle matches, I suppose we should consider things like Site Title, Site Logo, Site Tagline, Navigation, etc to be 'content'. Or perhaps there is a better way?

**Comment by kevin940726** [comment]
> but I am seeing results that I would consider to be unexpected. For example I have a Group containing a single Paragraph block, and when I shuffle I'm presented with Header and Footer patterns.

Yep! That's a known side effect that could happen as discussed in #44581.

> I suppose we should consider things like Site Title, Site Logo, Site Tagline, Navigation, etc to be 'content'. Or perhaps there is a better way?

This handcrafted list could properly work for core blocks but not for third-party blocks. I'm not sure if there's a better way though, at least not without introducing a new API. 🤔

**Comment by jameskoster** [comment]
I can't speak to the technical side, but for this feature to work it seems very important that the 'content' blocks remain consistent as you shuffle. Would it involve a new API to allow blocks to specify whether they are 'content' blocks?

**Comment by kevin940726** [comment]
There's already an API (`__experimentalRole`) to specify which attribute has the `content` role, which is what I'm using in this PR too. Right now, I identify blocks that have any `content` attribute as content blocks. So for instance, a Paragraph block is a content block because it has the `content` attribute which has `__experimentalRole: "content"`.

Upon shuffling, All the blocks and their attributes are changed except for the content attributes. That is, any attribute in the Paragraph block gets shuffled into what is defined in the pattern, except for the `content` attribute. This allows us to keep the user input (which is the only thing user can edit in `contentOnly` mode) when shuffling.

Is this the expected result you have in mind?

**Comment by jameskoster** [comment]
> Is this the expected result you have in mind?

Yeah how you described it sounds fine. So it seems we need to add `__experimentalRole: "content"` to the relevant attributes in blocks like Site Title, Site Tagline, Site Logo. That would 'fix' the 'issue' I described in https://github.com/WordPress/gutenberg/pull/45618#issuecomment-1308560597.

**Comment by talldan** [comment]
> So it seems we need to add __experimentalRole: "content" to the relevant attributes in blocks like Site Title, Site Tagline, Site Logo

Interesting. The technical problem is that those blocks [don't have attributes that can be marked with the content role](https://github.com/WordPress/gutenberg/blob/7147465adf207a3aa6717a882d504d33934702a9/packages/block-library/src/site-title/block.json#L10-L24). The content for those blocks is stored as part of the site itself.

This is at the very least a good discovery to make, even if it is a bit of a problem.

**Comment by youknowriad** [comment]
Yeah hard problem here, here are some ideas:

> Flatten all blocks and inner blocks in the patterns.

This sounded weird to me, while there are some blocks that need to be ignored while comparing (mostly style related blocks like "groups", "cover"....), some container blocks shouldn't be ignored like "quote", "list" blocks and I believe the new versions of these blocks don't have a "content" role either.

 > The technical problem is that those blocks [don't have attributes that can be marked with the content role](https://github.com/WordPress/gutenberg/blob/7147465adf207a3aa6717a882d504d33934702a9/packages/block-library/src/site-title/block.json#L10-L24). The content for those blocks is stored as part of the site itself.

This is also an interesting observation that may get us close to what should be the algorithm here :) 

----

So maybe the algorithm should be something like:

 1- Take the list of blocks and remove from the tree all container blocks that are just about layout: groups, covers, columns, spacer (not sure if there are others)
 2- Compare the remaining trees and the list of remaining blocks should match. The position of the blocks might differ but all the blocks should be available in the new pattern.
 3- Swap the pattern while retaining the "content" attributes (if they exist) in the matching blocks.

The problem I see in my algorithm is that if a pattern has "three" paragraphs, we would be able to shuffle it into a "four" paragraph pattern but we won't be able to get back to "three" anymore. So we could consider only matching exact number of filtered blocks. The problem is that we might end up with too few options. 

Also another problem is that if you can't transform from a query with "post-title" and "post-content" to a query with just "post-content". Makes me wonder if some blocks should be considered mandatory and others optional for example:

 - A query block is mandatory
 - All post blocks are optional
 - A quote block is mandatory
 - A button block (probably mandatory)
 - A paragraph block (probably mandatory / arguable)
 - A list block (I can't really tell)

--- 
Anyway, this is just some rumbling to try to get us into the right direction. No perfect solution here it seems.

**Comment by youknowriad** [comment]
The other thing that comes to my mind is that it seems that this is an AI ML problem. Maybe we can train a model manually using the pattern directory and apply it here. I'm not expert in ML but conceptually it seems like a good fit. cc @ellatrix

**Comment by talldan** [comment]
> The problem is that we might end up with too few options.

Yeah, I think this is a problem 🤔 The more accurate the system is, the fewer matches that can be presented, and actually that might make it less useful and less serendipitous.

It could be possible to loosen the rules and instead switch to a model where content loss is ok, but I think the UI would have to be different to a shuffle, and offer an easy way back to the original set of blocks (this might be a valid feature to provide anyway).

**Comment by jameskoster** [comment]
> The problem is that we might end up with too few options

I agree it would be a shame if there was only 2 patterns to shuffle between, but imo it's not a deal-breaker. I'd say that experience is significantly better when compared to shuffling between incompatible pattern configs. 

If we're worried about underwhelming, we might say "only allow shuffling when there are >_x_ matches".

It's also worth noting that it's a problem that will diminish over time after we connect to the pattern directory, and the directory itself grows.

**Comment by kevin940726** [comment]
In [b35142a](https://github.com/WordPress/gutenberg/pull/45618/commits/b35142a3681eae831a6775f41f0549dced25351f), I inline two functions `filterBlocksForShuffle` and `compareFilteredBlocks` for implementing the algorithm. The names are intentionally vague until we decide the path we want to take.

This commit roughly follows the algo @youknowriad shared in https://github.com/WordPress/gutenberg/pull/45618#issuecomment-1319651542. Instead of hardcoding the layout blocks, I use the `category` field to filter out those in the `design` category. This is not ideal either though, since it also won't work with third-party blocks. I wonder if there's something like `__experimentalRole: "layout"` for blockTypes?

This approach is obviously very precise though. Let's test it out and see if we can spot something we can improve.

**Comment by talldan** [comment]
> I wonder if there's something like __experimentalRole: "layout" for blockTypes?

@kevin940726 Possibly `hasBlockSupport( blockName, '__experimentalLayout' )`?

**Comment by kevin940726** [comment]
That could work! That will also match something like `navigation` and `post-content`, but it might be acceptable. Used it in [30845c6](https://github.com/WordPress/gutenberg/pull/45618/commits/30845c69f35e54e8cd7c6f24feecd08768a299ea).

**Comment by talldan** [comment]
It does feel like checking block supports is a better option, but at the same time I'd imagine the navigation block shouldn't be excluded (potentially its `ref` (selected menu) should carry over between patterns?).  I'm also not sure how the nav block is intended to work in `contentOnly` mode. I guess it would be uneditable. It's in a difficult gray area as a block.

**Comment by talldan** [comment]
Gave this a test. It works very well with some patterns. I tried inserting a basic footer (Default footer from TT3), and it shuffled between a few different options:

https://user-images.githubusercontent.com/677833/205582249-4fd4afe1-e4dc-4105-a18b-b5a61c0b6efd.mp4

As you can see, it doesn't consider the navigation block as content so it will use patterns that don't include it and may lose the selected menu.

The other thing I noticed is that I inserted this pattern (Two columns of text and title):
Screen Shot 2022-12-05 at 3 50 41 pm

It doesn't come wrapped in a group block by default, so I wrapped it myself and added the content locking attribute. When I clicked shuffle, I guess the only available pattern was the one I inserted and it removed the wrapping group block. It feels like the `contentOnly` wrapper should be preserved whenever shuffling, otherwise the 'Shuffle' button can unexpectedly disappear.

**Comment by kevin940726** [comment]
> It feels like the `contentOnly` wrapper should be preserved whenever shuffling, otherwise the 'Shuffle' button can unexpectedly disappear.

It is preserved, but only at the top level. As soon as we modify the pattern manually (wrap it with a group block in this case), we can no longer reference it back to the original pattern (performed by a naive algorithm in `areBlocksAlike`), hence we include the original pattern in the shuffle list. If the `contentOnly` attribute only applies to the group block then I suppose we can do something as you mentioned automatically? 🤔

**Comment by mtias** [comment]
Here's my take on this:

- For template part / semantic patterns we should not restrict things (all headers, footer, query blocks should be available for cycling) — the semantic role assumes the pairing and discarded blocks here are easy to retrieve so they don't need mapping (i.e. a site title would always be available).
- For arbitrary pattern sections we should be restrictive within a given category (i.e. cycle through "call to action" patterns, through "gallery" patterns, etc) while mapping content attributes and apply our logic on top of it.
  - We should try to hold on to discarded data until a block is unselected, so if you cycle through a loop you can get back to your possibly missed or discarded blocks.

Ultimately, I agree we should not develop too many custom heuristics and that pattern matching is something we could employ AI training to assist with.

**Comment by kevin940726** [comment]
@mtias Thanks for the feedback! Not sure if I understand it correctly, but there are some technical difficulties preventing us from implementing some of the details you mentioned.

1. There is currently no way to determine a pattern's semantic role after being inserted. We can't distinguish if it's a header, footer, text, or a combination of either.
2. There is also currently no API to determine if a block or an attribute is _essential_ to the pattern. For instance, a site title _might_ be essential to the pattern, and we want to keep it when cycling through different options.
3. After shuffling to a different pattern, the original information is lost. The whole blocks get replaced and the state is reset. This technically is possible though, we can store the state elsewhere semi-globally (possibly Redux store). We do want to make sure what the expection is though as this can get tricky with non-idiomatic code.

> Ultimately, I agree we should not develop too many custom heuristics and that pattern matching is something we could employ AI training to assist with.

Are we saying that maybe it's worth holding this for now until we research the AI possibilities? How do you imagine the implementation with AI would be? Do we implement the AI in the browser or in the users' WordPress instance? Or do we implement it in a WordPress.org API and call it from the client side?

**Comment by kevin940726** [comment]
I opened an alternative PR in #47594 with a slightly different approach: instead of shuffling on inserted patterns, we only allow shuffling on the pattern blocks. This way, we can keep the information of the pattern and shuffle to different patterns with the same `categories`.

**Pull request closed without merging**